File: ios/VisionCameraTranslator.swift
import Foundation
import VisionCamera
import MLKitVision
import MLKitTextRecognition
import MLKitTranslate

public class VisionCameraTranslator: FrameProcessorPlugin {
    private var models: [String] = []
    private var isDownloaded: Bool = false
    private let localModels = ModelManager.modelManager().downloadedTranslateModels
    private var translator: Translator
    private var translatorOptions: TranslatorOptions
    private var text = ""
    private let conditions = ModelDownloadConditions(
        allowsCellularAccess: false,
        allowsBackgroundDownloading: true
    )
    private var from = ""
    private var to = ""
    private var textRecognizer = TextRecognizer.textRecognizer(options: TextRecognizerOptions())

    public override init(proxy: VisionCameraProxyHolder, options: [AnyHashable: Any]! = [:]) {
        let from = options["from"]  as? String ?? "en"
        let to = options["to"]  as? String ?? "de"

        self.from  = from
        self.to  = to

        let sourceLanguage = TranslateLanguage(from: self.from) ?? .english
        let targetLanguage = TranslateLanguage(from: self.to) ?? .german
        self.translatorOptions = TranslatorOptions(sourceLanguage: sourceLanguage, targetLanguage: targetLanguage)
        self.translator = Translator.translator(options: translatorOptions)
        super.init(proxy: proxy, options: options)
        downloadModel()
    }

    public override func callback(_ frame: Frame, withArguments arguments: [AnyHashable: Any]?) -> Any {
        let buffer = frame.buffer
        let image = VisionImage(buffer: buffer)

        let orientation = getOrientation(
               orientation: frame.orientation
             )
        image.orientation = orientation
        let resultText = recognizeText(image: image)!.text
        if isDownloaded && !resultText.isEmpty {
            self.translateText(text:resultText){translatedText in
                self.text = translatedText!
            }
        }
        if resultText.isEmpty {
            self.text = ""
        }
        return self.text
    }
    private func downloadModel(){
        localModels.forEach { language in
            models.append(language.language.rawValue)
            if models.contains(from) && models.contains(to) {
                self.isDownloaded = true
            }
        }
        if !isDownloaded {
            translator.downloadModelIfNeeded(with: conditions) { error in
                guard error == nil else {
                    self.isDownloaded = false
                        return
                }
                self.isDownloaded = true
            }
        }
    }

    private func recognizeText(image:VisionImage)-> Text?{
        do {
            let result = try self.textRecognizer.results(in: image)
            return result
        } catch {
            print("Failed to recognize text: \(error.localizedDescription).")
            return nil
        }

    }
    private func getOrientation(orientation: UIImage.Orientation) -> UIImage.Orientation {
        switch orientation {
        case .up:
          return .up
        case .left:
          return .right
        case .down:
          return .down
        case .right:
          return .left
        default:
          return .up
        }
    }

    private func translateText(text: String, completion: @escaping (String?) -> Void) {
        if !isDownloaded {
            print("Translation model is not downloaded.")
            completion(nil)
            return
        }
        translator.translate(text) { translatedText, error in
            guard error == nil, let translatedText = translatedText else {
                print("Translation failed: \(error?.localizedDescription ?? "Unknown error").")
                completion(nil)
                return
            }
            completion(translatedText)
        }
    }
}

File: android/src/main/java/com/visioncameratextrecognition/VisionCameraTranslator.kt
package com.visioncameratextrecognition

import android.media.Image
import android.util.Log
import android.util.SparseIntArray
import android.view.Surface
import com.google.android.gms.tasks.Task
import com.google.android.gms.tasks.Tasks
import com.google.mlkit.common.model.DownloadConditions
import com.google.mlkit.common.model.RemoteModelManager
import com.google.mlkit.nl.translate.TranslateLanguage
import com.google.mlkit.nl.translate.TranslateRemoteModel
import com.google.mlkit.nl.translate.Translation
import com.google.mlkit.nl.translate.TranslatorOptions
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.Text
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import com.mrousavy.camera.core.types.Orientation
import com.mrousavy.camera.frameprocessors.Frame
import com.mrousavy.camera.frameprocessors.FrameProcessorPlugin
import com.mrousavy.camera.frameprocessors.VisionCameraProxy

class VisionCameraTranslator(proxy: VisionCameraProxy, options: Map<String, Any>?) :
    FrameProcessorPlugin() {

    private val conditions = DownloadConditions.Builder().requireWifi().build()
    private var translatorOptions = TranslatorOptions.Builder()
        .setSourceLanguage(TranslateLanguage.ENGLISH)
        .setTargetLanguage(TranslateLanguage.GERMAN)
        .build()
    private val translator = Translation.getClient(translatorOptions)
    private var recognizer = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)
    private val ORIENTATIONS = SparseIntArray()
    private var from = ""
    private var to = ""
    private var models = emptyArray<String>()
    private var isDownloaded: Boolean = false
    private val modelManager = RemoteModelManager.getInstance()
    private var translatedText = ""
    init {
        ORIENTATIONS.append(Surface.ROTATION_0, 0)
        ORIENTATIONS.append(Surface.ROTATION_90, 90)
        ORIENTATIONS.append(Surface.ROTATION_180, 180)
        ORIENTATIONS.append(Surface.ROTATION_270, 270)
        this.from = options?.get("from") as? String ?: ""
        this.to = options?.get("to") as? String ?: ""
        val sourceLanguage = translateLanguage(from)
        val targetLanguage = translateLanguage(to)
        if (sourceLanguage != null && targetLanguage != null) {
            translatorOptions = TranslatorOptions.Builder()
                .setSourceLanguage(sourceLanguage)
                .setTargetLanguage(targetLanguage)
                .build()
        } else {
            println("Invalid language strings provided for translation.")
        }
        downloadModel()
    }

    override fun callback(frame: Frame, params: MutableMap<String, Any>?) : Any? {
        val mediaImage: Image = frame.image
        val rotation = getOrientation(frame.orientation)
        val image = InputImage.fromMediaImage(mediaImage, rotation)
        val task: Task<Text> = recognizer.process(image)
        try {
            val resultText: String = Tasks.await<Text>(task).text
            Log.d("TEXT", "TEXT $resultText")
            if (this.isDownloaded && resultText.isNotEmpty()) {
                this.translateText(resultText)
            }
            if (resultText.isEmpty()){
                this.translatedText = ""
            }
            return this.translatedText
        }catch (e:Exception){
            Log.e("ERROR", "$e")
            return null
        }
    }

    private fun translateText(text: String) {
      //  this.translatedText = ""
        translator.translate(text)
            .addOnSuccessListener {
                this.translatedText = it
            }
            .addOnFailureListener {
                Log.e("ERROR", "$it")
                this.translatedText = ""
            }
    }


    private fun getOrientation(
        orientation: Orientation
    ): Int {
        return when (orientation) {
            Orientation.PORTRAIT -> 0
            Orientation.LANDSCAPE_LEFT -> 90
            Orientation.PORTRAIT_UPSIDE_DOWN -> 180
            Orientation.LANDSCAPE_RIGHT -> 270
        }
    }
    private fun downloadModel(){
        modelManager.getDownloadedModels(TranslateRemoteModel::class.java)
            .addOnSuccessListener {
                it.forEach() { language ->
                    models.plus(language.language)
                }
                if (models.contains(from) && models.contains(to)) {
                    isDownloaded = true
                }
            }
            .addOnFailureListener {
                this.isDownloaded = false
            }

        if (!isDownloaded){
            translator.downloadModelIfNeeded(conditions)
                .addOnSuccessListener {
                    this.isDownloaded = true
                }
                .addOnFailureListener {
                    this.isDownloaded = false
                }
        }
    }
}
